packages/engine/src/lib/action/branch-action-handler.ts

```diff
@@ -28,52 +35,72 @@ export class BranchActionHandler extends BaseActionHandler<BranchAction> {
     ancestors: [string, number][]
   ): Promise<StepOutput> {
     const resolvedInput = await this.variableService.resolve(
-      this.action.settings,
+      this.currentAction.settings,
       executionState
     );
 
     const stepOutput = new BranchStepOutput();
     stepOutput.input = await this.variableService.resolve(
-      this.action.settings,
+      this.currentAction.settings,
       executionState,
       true
     );
 
     try {
-      const condition = evaluateConditions(resolvedInput.conditions);
+      const condition = this.resumeStepMetadata
+        ? this.resumeStepMetadata.conditionEvaluation
+        : evaluateConditions(resolvedInput.conditions);
+
       stepOutput.output = {
         condition: condition,
-      };
-      executionState.insertStep(stepOutput, this.action.name, ancestors);
-      const executor = new FlowExecutor({ executionState });
-      if (condition) {
-        if (this.onSuccessAction) {
-          await executor.iterateFlow(
-            this.onSuccessAction,
-            ancestors
-          );
-        }
-      } else {
-        if (this.onFailureAction) {
-          await executor.iterateFlow(
-            this.onFailureAction,
-            ancestors
-          );
-        }
       }
+
+      executionState.insertStep(stepOutput, this.currentAction.name, ancestors);
+
+      const firstStep = condition
+        ? this.onSuccessAction
+        : this.onFailureAction
+
+      if (isNil(firstStep)) {
```

Comment:
OnSuccessAction / onFailureAction is optional field, so throwing an error would make the flow stop.